### ISSUE TITLE
Preventing new features to be integrated into 3.0

### ISSUE THREAD

**Issue by vaibhav-kaushal** [qcubed/framework#801]
Hello all (especially @spekary ),

I think we have enough features and enhancements into QCubed 3.0 as of now. If there are minor enhancements, we can push them into 3.1 (or 3.0.1, if they are small enough). I have a large project which is on Beta 3.0 and we are trying to see what all breaks. So far not much has but there are a few bugs which I have found. Some are fixed, others are not. I will be creating the pull requests for the fix soon. However it becomes problematic if new features are coming in constantly. 

I believe we have added enough for now and we should now focus on fixing bugs and releasing version 3.0. We are more than 1.5 years late already. :stuck_out_tongue:

**Comment by spekary** [comment]
Understood. However, the reason it is taking so long is that there simply is a lot to do. We put some goals down for this release, and we are almost there:
- HTML 5 compliant. (Hence, the QDataGrid2). I think it would be problematic to release a new major version with so much of the codegen and framework relying on QDataGrid, when QDatGrid is slow, uses EVAL, and uses deprecated HTML that is not even supported in HTML5. This is now in.
- Codegen using the model connector editor. This is now supported for QDataGrid2 and QDatatables.
- Being able to create a Bootstrap codegenned website using the Bootstrap plugin and the model connector editor. This is what I am working on now, and what the feature requests I have made are for. The features are very minor enhancements to allow the bootstrap plugin to do some things it needs to do.

One final thing before we release is to take a step back and ask what else needs to happen with a major release. There is one thing that concerns me still.  The names of some things are problematic for newcomers. In particular, QColumn and QTable are codegen things only and refer to the database. QSimpleTable was born out of a desire to simplify the QDataGrid, but QTable was taken by codegen. And QDataGrid is obsolete. I think a global rename is needed:
  - QTable -> QDbTable
  - QColumn -> QDbColumn
  - QSimpleTable -> QTable
  - QAbstractSimpleTableColumn -> QTableColumnBase
  - And the other columns renamed in similar ways
  - QDataGrid -> QDataGridOld
  - QDataGrid2 -> QDataGrid

QDataGrid examples need to be rewritten to use the new QDataGrid.

And then of course, someone needs to work on the website to do the release.

**Comment by kukrik** [comment]
I look forward to the final release. Success to @spekary :+1:! Unfortunately, I am still busy with other work. But I am willing to commit to testing, If you know me...

**Comment by kukrik** [comment]
To @vaibhav-kaushal, It is the largest restructuring, of course, our support @spekary work...

**Comment by vaibhav-kaushal** [comment]
@spekary I agree. Those are some real good changes and I think they are difficult to undertake too. Thanks for taking the initiative and putting in the effort. However, if it were possible, could we not put some of them into 3.1? I am having a major challenge on my end getting things in order. 

Also, in case you need help with something, please let me know. Though I am very busy, QCubed is of huge importance to me (and my company) as well and I would be ready to chip in and help with what I can.

**Comment by vaibhav-kaushal** [comment]
@kukrik I admire his work as well. There is no doubt about that :)

**Comment by spekary** [comment]
I think that is fine. We can lock in the current one. Submit the bug fixes you have. I don't currently have any more. We can push name changes to next version. I am concerned that with the name changes and other 3.1 planned items (like namespacing), documentation will need to change significantly. 

BTW, I have done some testing with HHVM, and all seems to work fine there. Not sure about PHP7 though.

I will be working on the Bootstrap plugin in the near term, so I don't think I need any help per se. The next real work is to walk through all the examples to confirm they still work (Should we change examples to use QDataGrid2?), and then we should update the website. Did someone volunteer to update the drupal website? Should we move to something else? Should we use a github website as the front end? We of course would still need a demo site.

**Comment by vaibhav-kaushal** [comment]
@spekary I have made the pull request ( #802 ). 

About the change in documentation - yes we will have to update it. Slowly with time, it would happen.

I have not tested it yet on HHVM or PHP7. I would rather wait for the final release. HHVM is faster but PHP7's engine is competitive of that of HHVM. So, I am not very eager about HHVM for now (unless there are magical super powers :wink: ). 

Once the release is done, I have planned for documenting the framework with examples, tips, tricks etc. So that would take time. Meanwhile, I would love to work on Bootstrap plugin as well (at least I can provide you some input because most of what I am doing makes use of Bootstrap). Let's see about it then.

**Comment by koopa** [comment]
I would also like to help with some of the work related to the Connectors and Datagrids.

I am using

    "qcubed/framework": "dev-beta-3.0"


`framework\assets\php\examples\other\datagrid_connectors.php` seemed like a good point to start, but there I saw that neither `ProjectDataGrid` nor `AddConnectedColumn` are found in the generated code.
    
Also, [in the wiki Upgrade page](https://github.com/qcubed/framework/wiki/Upgrade3) `AddConnectedColumn` shows up, but as already mentioned I could not find any references.

I am going to try and understand what is going on exactly. Has anyone already done some work on `AddConnectedColumn` and would be willing to send me a pull request? And maybe answer a few questions I might have.

**Comment by koopa** [comment]
Clarification:
The class can actually be found by the AutoLoader via 
`QApplicationBase::$ClassFile['projectdatagrid'] =  _MODEL_CONNECTOR__ . '/ProjectDataGrid.class.php';` but the file `ProjectDataGrid.class.php` it is trying to include does not exist.
`

**Comment by spekary** [comment]
I removed AddConnectedColumn recently. It is now AddNodeColumn. Looks like I neglected to update the examples to reflect this change. This was part of a simplification of the generation of datagrids, and removing dependency on QDataGrid to allow plugin datagrids. The basic codegen should now be creating subclasses of QDataGrid2. The old QDataGrid had quite a few problems: reliance on HTML4 only attributes, reliance on PHP Eval, generating a QControl for every row of the table.

This is all fairly new work, so not surprised you found issues. If you can help clean up the examples, that would be great.

**Comment by koopa** [comment]
Alright, I will try getting it to work. It seems like instead of ProjectDataGrid I now should use ProjectList, is that correct? Also I still couldn't find any references to AddNodeColumn, is that possibly not pushed yet or am I missing something?

**Comment by spekary** [comment]
Sorry, CreateNodeColumn. Just search for NodeColumn. And yes, ProjectList is the new thing now.

**Comment by koopa** [comment]
Thanks for the Information. I was able to successfully convert some of my ListForms that used old MetaDatagrids into the new 3.0 format.

Since I have literally hundreds of forms where I need to refactor the datagrids I wrote a small script to eliminate some of the manual replacement hassle. If I run into new problems I will keep updating the script to add new refactoring rules.
My codegen conventions should be congruent with the defaults AFAIK. So maybe it will even work to convert the examples with this - I haven't tried yet.

```bash
#!/bin/bash
#
# qc form migration - in-place replacements
# 
[[ -z $1 ]] && echo "Argument required: $0 some_form.php [[another_panel.php] ...]" && exit 1;

echo "Processing: $1"

search=(QMetaControl MetaControl MetaAddColumn MetaAddEditLinkColumn MetaAddEditProxyColumn MetaDataBinder)
replace=(QModelConnector Connector CreateNodeColumn AddEditLinkColumn AddEditProxyColumn ConnectorDataBinder)
for i in ${!search[@]}; do
    echo sed -r -i -e 's/'${search[$i]}'/'${replace[$i]}'/g' "$1"
    sed -r -i -e 's/'${search[$i]}'/'${replace[$i]}'/g' "$1"
done


echo sed -i -r -e 's/dtg([A-Z][a-zA-Z]*)s->CreateNodeColumn\((['"'"'"])([^'"'"'"]+)\2\)/dtg\1s->CreateNodeColumn(\2\3\2, QQN::\1()->\3)/g' "$1"
sed -i -r -e 's/dtg([A-Z][a-zA-Z]*)s->CreateNodeColumn\((['"'"'"])([^'"'"'"]+)\2\)/dtg\1s->CreateNodeColumn(\2\3\2, QQN::\1()->\3)/g' "$1"
echo sed -i -r -e 's/dtg([A-Z][a-zA-Z]*)s->CreateNodeColumn\(QQN::\1\(\)->([A-Za-z]*)\)/dtg\1s->CreateNodeColumn("\2", QQN::\1()->\2)/g' "$1"
sed -i -r -e 's/dtg([A-Z][a-zA-Z]*)s->CreateNodeColumn\(QQN::\1\(\)->([A-Za-z]*)\)/dtg\1s->CreateNodeColumn("\2", QQN::\1()->\2)/g' "$1"

echo "Done. Diff and check changes."
```


Relevant excerpt of a diff afterwards:
```
-    $this->dtgReviewPrototypes->MetaAddColumn(QQN::ReviewPrototype()->Responsible);
+    $this->dtgReviewPrototypes->CreateNodeColumn("Responsible", QQN::ReviewPrototype()->Responsible);
-    $this->dtgReviewPrototypes->MetaAddColumn('Name');
+    $this->dtgReviewPrototypes->CreateNodeColumn('Name', QQN::ReviewPrototype()->Name);
```

**Comment by koopa** [comment]
I created a pull request for the datagrid_connectors example, it should work now.
Not sure about the AdditionalCondition/-Clauses thing, I just tried to reproduce the original example as simple as possible. Also the description needs to be changed a bit in the tpl file, since a few details have changed. Sorry about the white space changes that I missed until after the push.

**Comment by koopa** [comment]
One more thing: What is the new way for the deprecated `AddEditProxyColumn`?
This also needs to be done in the example, but I wasn't exactly sure how.
Also by default there is no EditProxyColumn needed anymore with the clickable rows anyway, but it would still couldn't harm as an example.

**Comment by spekary** [comment]
This is currently not implemented. I recently changed how proxies work to be more efficient, so it could be added now as an option.

**Comment by koopa** [comment]
I also created a new node creation function `CreateNodeCallableColumn` and corresponding class `QSimpleTableNodeCallableColumn` which combines both the goodness of automatic qq-node mapping (for sorting, etc.) while allowing custom cell render methods. 
Also abstracted the callable related code into a `trait CallableColumn` and added some callback normalizing functions to allow both [[cb], [args, ...]] and [cb, args, ...] style callback arguments.

Still in the progress of migrating quite a large qcubed 2.0 installation to 3.0, so you might need to give me a week or two until I get to merging the stuff back into a pristine state for the beta-3.0 branch. Just informing to prevent work from being done twice - in case somebody intended to do exactly those things.

**Comment by spekary** [comment]
See objPostCallback in QSimpleTableColumn. There is already a means to add a custom render method to any column.

**Comment by vaibhav-kaushal** [comment]
I was thinking about allowing plugins to add more features and custom methods into the codegen process. That way, everyone using the framework can create and use their own templates with whatever functionality they want. I believe, in such a case, we would have to modify the process a little bit and have the provision to use templates as a plugin while keeping the default ones in the framework. But that would be in 3.1. 

Will be making a new issue for that.

**Comment by spekary** [comment]
This capability is already in 3.0. to use it, edit the QCodeGen constructor and set the static TemplatePaths variable like this:
```
			static::$TemplatePaths = array (
				__QCUBED_CORE__ . '/codegen/templates/',
				__DOCROOT__ . __SUBDIRECTORY__ . '/lib/templates/',
				__QCUBED__ . '/codegen/templates/'
```
A file later in the array will override a file earlier in the array.

**Comment by vaibhav-kaushal** [comment]
Continuing the discussion about this feature in #815 .
